### PR TITLE
Sol anonymous events

### PR DESCRIPTION
**Pull request by LianaHus**

### CODE REVIEW

File: libsolidity/AST.h [hunk 1]
```diff
@@ -556,14 +556,16 @@ class EventDefinition: public Declaration, public VariableScope, public Document
 	EventDefinition(SourceLocation const& _location,
 					ASTPointer<ASTString> const& _name,
 					ASTPointer<ASTString> const& _documentation,
-					ASTPointer<ParameterList> const& _parameters):
-		Declaration(_location, _name), Documented(_documentation), m_parameters(_parameters) {}
+					ASTPointer<ParameterList> const& _parameters,
```
Comment (chriseth) [inline libsolidity/AST.h]: Please change to "new" coding style in each line you make a change.

File: libsolidity/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -538,9 +538,12 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
 				}
-			m_context << u256(h256::Arith(dev::sha3(function.getCanonicalSignature(event.getName()))));
-			++numIndexed;
-			solAssert(numIndexed <= 4, "Too many indexed arguments.");
+			if (!event.IsAnonymous())
+			{
+				m_context << u256(h256::Arith(dev::sha3(function.getCanonicalSignature(event.getName()))));
+				++numIndexed;
+			}
+			solAssert(numIndexed <= 4 - (event.IsAnonymous() ? 1 : 0), "Too many indexed arguments.");
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: It should be 4 in both cases.

File: test/SolidityEndToEndTest.cpp [hunk 3]
```diff
@@ -2179,6 +2179,21 @@ BOOST_AUTO_TEST_CASE(event_no_arguments)
 	BOOST_CHECK_EQUAL(m_logs[0].topics[0], dev::sha3(string("Deposit()")));
 }
 
+BOOST_AUTO_TEST_CASE(event_anonymous)
+{
+	char const* sourceCode = R"(
+		contract ClientReceipt {
+			event anonymous Deposit();
```
Comment (chriseth) [inline test/SolidityEndToEndTest.cpp]: To be more in line with function declaration, I think it is better to put anonymous just in front of the `;`.

File: libsolidity/AST.h (hunk 1 above)
Comment (LianaHus) [inline libsolidity/AST.h]: ok

File: test/SolidityEndToEndTest.cpp (hunk 3 above)
Comment (LianaHus) [inline test/SolidityEndToEndTest.cpp]: I'll change it

**Comment by chriseth** [comment]
Please add another test that checks that anonymous events with user-supplied topics work as expected.

File: libsolidity/AST.h [hunk 4]
```diff
@@ -574,6 +581,7 @@ class EventDefinition: public Declaration, public VariableScope, public Document
 
 private:
 	ASTPointer<ParameterList> m_parameters;
+	bool m_anonymous;
```
Comment (chfast) [inline libsolidity/AST.h]: For clarity and safty you can init the field in declaration: `bool m_anonymous = false`.

Comment (LianaHus) [inline libsolidity/AST.h]: added :)

**Comment by LianaHus** [comment]
rebased

**Comment by chfast** [comment]
LGTM

**Pull request merged**